### PR TITLE
Add LLVM intrinsic functions

### LINKED ISSUE #30: Implement nback.llvm support for calling intrinsic functions
We can call intrinsics using a `call` instruction like other functions. But we also need to create a `Function` that refers to them. The nback.llvm module should take care of emitting the appropriate declaration.

LLVM C API has:

* LLVMGetIntrinsicDeclaration
* LLVMLookupIntrinsicID

We may need two types of Function.

This is needed in order to be able to use #24 to support arithmetic with overflow.

### PR DESCRIPTION
**Pull request by heshanpadmasiri**
## Purpose
Implement nback.llvm support for calling intrinsic functions
Resolve #30 

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: compiler/modules/nback.llvm/llvm.bal [hunk 1]
```diff
@@ -87,6 +89,29 @@ public class Module {
...
+                self.functionDeclns.push(fn);
+                return fn;
+            }
+            "smul.with.overflow.i64" => {
+                FunctionDecl fn = new ("llvm.smul.with.overflow.i64", {returnType: overflowArithmeticReturnType, paramTypes: ["i64", "i64"]});
+                self.functionDeclns.push(fn);
+                return fn;
+            }
+        }
+        panic error ("Unknown intrinsic function");
```
Comment (jclark) [inline compiler/modules/nback.llvm/llvm.bal]: This is unreachable, so use `err:unreached`

File: compiler/modules/nback.llvm/llvm.bal [hunk 2]
```diff
@@ -87,6 +89,29 @@ public class Module {
         return fn;
     }
 
+    // Corresponds to LLVMGetIntrinsicDeclaration
+    public function getIntrinsicDeclaration(IntrinsicFunctionName name) returns FunctionDecl {
```
Comment (jclark) [inline compiler/modules/nback.llvm/llvm.bal:94]: The logic should add the intrinsic to the list of declarations only once, even if it is called multiple times. You will need a map or a table to manage this.

File: compiler/modules/nback.llvm/llvm.bal [hunk 3]
```diff
@@ -72,10 +72,12 @@ public readonly class PointerValue {
 public function constInt(IntType ty, int val) returns Value {
     return new Value(ty, val.toString());
 }
+public type IntrinsicFunctionName "sadd.with.overflow.i64"|"ssub.with.overflow.i64"|"smul.with.overflow.i64";
 
 # Corresponds to llvm::Module class
 public class Module {
     private final FunctionDefn[] functions = [];
+    private final FunctionDecl[] functionDeclns = [];
```
Comment (jclark) [inline compiler/modules/nback.llvm/llvm.bal]: functionDecls (not n)

File: compiler/modules/nback.llvm/llvm.bal [hunk 4]
```diff
@@ -87,6 +89,29 @@ public class Module {
         return fn;
     }
 
+    // Corresponds to LLVMGetIntrinsicDeclaration
+    public function getIntrinsicDeclaration(IntrinsicFunctionName name) returns FunctionDecl {
+        StructType overflowArithmeticReturnType = structType(["i64", "i1"]);
+        match name {
```
Comment (jclark) [inline compiler/modules/nback.llvm/llvm.bal:101]: Too much repeated code here

File: compiler/modules/nback.llvm/llvm.bal [hunk 5]
```diff
@@ -102,6 +127,9 @@ public class Module {
     }
 
     function output(Output out) {
+        foreach var functionDcln in self.functionDeclns {
```
Comment (jclark) [inline compiler/modules/nback.llvm/llvm.bal]: Abbreviate "declaration" consistently as "decl"

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/nback.llvm/llvm.bal (hunk 2 above)
Comment (heshanpadmasiri) [reply compiler/modules/nback.llvm/llvm.bal:94]: Fixed

File: compiler/modules/nback.llvm/llvm.bal (hunk 4 above)
Comment (heshanpadmasiri) [reply compiler/modules/nback.llvm/llvm.bal:101]: Fixed

File: compiler/modules/nback.llvm/llvm.bal (hunk 1 above)
Comment (heshanpadmasiri) [reply compiler/modules/nback.llvm/llvm.bal]: Fixed

File: compiler/modules/nback.llvm/llvm.bal (hunk 3 above)
Comment (heshanpadmasiri) [reply compiler/modules/nback.llvm/llvm.bal]: Fixed

File: compiler/modules/nback.llvm/llvm.bal (hunk 5 above)
Comment (heshanpadmasiri) [reply compiler/modules/nback.llvm/llvm.bal]: Fixed

File: compiler/modules/nback.llvm/llvm.bal [hunk 6]
```diff
@@ -125,6 +158,23 @@ public class FunctionDecl {
         self.functionName = functionName;
         self.functionType = functionType;
     }
+
+    function output(Output out) {
```
Comment (jclark) [inline compiler/modules/nback.llvm/llvm.bal:161]: This function has a lot of common code with `header`. Please factor out the common bits into a separate function.

**Review by jclark — COMMENTED** [review]
Can you factor out the common code for output of function declarations and the header of function definitions?

File: compiler/modules/nback.llvm/llvm.bal (hunk 6 above)
Comment (heshanpadmasiri) [reply compiler/modules/nback.llvm/llvm.bal:161]: Fixed

**Comment by heshanpadmasiri** [comment]
> Can you factor out the common code for output of function declarations and the header of function definitions?

Fixed by adding new function `functionHeader`

**Pull request merged**